secure_software_development_fundamentals.md

```diff
@@ -2411,17 +2431,121 @@ Of course, like any technique, if you use it wrongly then it won’t be secure.
+  Users can later use the separate `PQexecPrepared` call
+  to provide the placeholder parameter values and execute the resulting
+  command.
 
-Properly using prepared statement libraries makes it much easier to write secure code. In addition, they typically make code easier to read, automatically handle the variations between how databases escape things, and sometimes they are faster than doing metacharacter escapes yourself.
+The [OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html) and [Bobby Tables website](https://bobby-tables.com/) provide examples for a variety of ecosystems.
+
+
+##### Stored Procedures
+
+Many database systems support "stored procedures", that is,
+procedures embedded in the database itself.
+If you use stored procedures, and commands (such as queries)
+are dynamically constructed in them, then you can again have
+SQL injection vulnerabilities.
+
+Again, the best solution is usually to use a parameterized query mechanism
+when creating the dynamic command (e.g., the SQL) inside the
+stored procedure. Be careful; in some systems using them correctly
+can be a little tricky.
+
+For more information on using parameterized queries
+in stored procedures, see your library's documentation, the
+[OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html), and the
+[OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
+
+##### When Parameterized Interfaces Won't Work
+
+In some situations parameterized statements (including
+prepared statements) will *not* work.
+Many parameterized interface APIs only allow replacing SQL values, so
+they do not allow varying information such as the names of tables, the names
+of columns, or the sort order direction.
+
+In those cases you may need to create a query by concatenating data.
+In those cases, make *sure* you carefully validate the data (using an allowlist)
+so only specific and safe values are allowed.
+Often the allowlist is a short list of permitted values, or at most
+a simple expression that only allows ASCII letters and digits.
+A risk with this approach is that if the validation
+is ever skipped (e.g., after some code change),
+the system may be become extremely vulnerable.
+
+The [OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
+has a nice suggestion: if you *must* do this concatenation,
+make the user input *different* from what you will actually use, and
+in the program map the user input to the values you will use.
+If user input doesn't have a valid mapping, reject the input.
+This increases the probability that a change to the program
+will not result in unvalidated input being concatenated into a query,
+or that the problem will be detected before shipping.
+
```

Comment:
This is important and common enough that I think we should give a concrete example -- feel free to change:

```suggestion
For example, in Python, if you need to write to a user-provided table name, you can do the following:

~~~~python
    table_name_untrusted = request.get("table_name")  # This is untrusted, don't put this directly in the query!
    table_name_map = {"table1": "db.table1", "table2": "db.table2"}
    table_name = table_name_map[table_name_untrusted]
    con = sqlite3.connect(...)
    cur = con.cursor()
    cur.execute(f"insert into {table_name}(d, ts) values (?, ?)", (today, now)) # This is safe because we know that table_name can only take trusted values from table_name_map
~~~~
```